### ISSUE TITLE
nn.Embedding can cause VRAM leak with variable length input

### ISSUE THREAD

**Issue by threewayhandshake** [pytorch/pytorch#122781]
### 🐛 Describe the bug

First I encountered unexpected increasing memory usage issue when I changed input for my model from fixed length to variable length, despite of max_length is the same.
Then I tried some tests and found out the cause is nn.Embedding apparently.
Finally I created a reproduction code and got following results.

When input length is variable, memory usage increased progressively.

```
SEQ_LEN_RANGE=(0, 1024) BATCH_SIZE=1024 VOCAB_SIZE=1000 DIM=256
i=1     max_vram_usage=506MB    historical_max_seq_len=484
i=2     max_vram_usage=1198MB   historical_max_seq_len=691
i=5     max_vram_usage=2004MB   historical_max_seq_len=805
i=10    max_vram_usage=2874MB   historical_max_seq_len=869
i=11    max_vram_usage=3706MB   historical_max_seq_len=869
i=13    max_vram_usage=4602MB   historical_max_seq_len=896
i=20    max_vram_usage=5528MB   historical_max_seq_len=925
i=36    max_vram_usage=6532MB   historical_max_seq_len=1004
i=52    max_vram_usage=6534MB   historical_max_seq_len=1004
i=106   max_vram_usage=7542MB   historical_max_seq_len=1008
i=206   max_vram_usage=8560MB   historical_max_seq_len=1017
i=227   max_vram_usage=9580MB   historical_max_seq_len=1020
i=294   max_vram_usage=10602MB  historical_max_seq_len=1022
i=365   max_vram_usage=11626MB  historical_max_seq_len=1023
[current i=1482, vram_usage=11626MB]
```

But when input length is fixed, memory usage did not increased.

```
SEQ_LEN_RANGE=(1024, 1024)      BATCH_SIZE=1024 VOCAB_SIZE=1000 DIM=256
i=1     max_vram_usage=1046MB   historical_max_seq_len=1024
i=2     max_vram_usage=2070MB   historical_max_seq_len=1024
[current i=223, vram_usage=2070MB]
```

The reproduction code is bellow.

```python
import torch
import torch.nn as nn
import random
import sys

DEVICE = 'cuda'
SEQ_LEN_RANGE = (0, 1024)
BATCH_SIZE = 1024
VOCAB_SIZE = 1000
DIM = 256

emb = nn.Embedding(VOCAB_SIZE, DIM, device=DEVICE)

i = 0
max_vram_usage = 0
historical_max_seq_len = 0
print(f'{SEQ_LEN_RANGE=}\t{BATCH_SIZE=}\t{VOCAB_SIZE=}\t{DIM=}')
while True:
    i += 1
    ### create random input with random length ###
    batch_max_seq_len = random.randint(SEQ_LEN_RANGE[0], SEQ_LEN_RANGE[1])
    seqs = []
    for _ in range(BATCH_SIZE):
        seq = [random.randrange(VOCAB_SIZE) for _ in range(batch_max_seq_len)]
        seqs.append(seq)
    input_ids = torch.tensor(seqs, dtype=torch.int32, device=DEVICE)
    ### forward ###
    x = emb(input_ids)
    ### check current vram usage and log if increased ###
    vram_usage = torch.cuda.max_memory_reserved() // 1024 // 1024
    torch.cuda.reset_peak_memory_stats()
    historical_max_seq_len = max(historical_max_seq_len, batch_max_seq_len)
    if max_vram_usage < vram_usage:
        max_vram_usage = vram_usage
        print('\r\033[K'f'{i=}\t{max_vram_usage=}MB\t{historical_max_seq_len=}')
    ### show progress ###
    print('\r\033[K'f'[current {i=}, {vram_usage=}MB]', end='', file=sys.stderr, flush=True)
```

### Versions

Collecting environment information...
PyTorch version: 2.2.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Home
GCC version: (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:26:36) [MSC v.1937 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: 8.0.60
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Ti
Nvidia driver version: 546.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=3696
DeviceID=CPU0
Family=198
L2CacheSize=1536
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=3696
Name=Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.2.0+cu121
[conda] Could not collect

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by threewayhandshake** [comment]
I got similar results using `torch.zeros(BATCH_SIZE, batch_max_seq_len, DIM, device=DEVICE)` instead of `emb(input_ids)` so it may not be caused by nn.Embedding essentially.

**Comment by threewayhandshake** [comment]
I was able to reproduce the issue also in Colab environment.
It does not seem to be platform dependent.

**Comment by mikaylagawarecki** [comment]
I can reproduce this. The output of the script is as such for me 

```
SEQ_LEN_RANGE=(0, 1024) BATCH_SIZE=1024 VOCAB_SIZE=1000 DIM=256
i=1     max_vram_usage=676MB    historical_max_seq_len=653
i=2     max_vram_usage=1120MB   historical_max_seq_len=653
i=5     max_vram_usage=2136MB   historical_max_seq_len=1016
i=20    max_vram_usage=3024MB   historical_max_seq_len=1016
i=63    max_vram_usage=3026MB   historical_max_seq_len=1016
i=86    max_vram_usage=4022MB   historical_max_seq_len=1016
i=220   max_vram_usage=5046MB   historical_max_seq_len=1024
[current i=2680, vram_usage=5046MB]
```

and for fixed input length

```
SEQ_LEN_RANGE=(1024, 1024)      BATCH_SIZE=1024 VOCAB_SIZE=1000 DIM=256
i=1     max_vram_usage=1046MB   historical_max_seq_len=1024
i=2     max_vram_usage=2070MB   historical_max_seq_len=1024
[current i=289, vram_usage=2070MB]
```

Further when I comment out the `x = emb(input_ids)` line, the behavior goes away so we can indeed isolate this to the `nn.Embedding` for variable length sequences.

Marking needs research regarding whether this behavior is expected

**Comment by mikaylagawarecki** [comment]
@threewayhandshake The metric that we want to check here is [`torch.cuda.max_memory_allocated()`](https://pytorch.org/docs/stable/generated/torch.cuda.max_memory_allocated.html#torch.cuda.max_memory_allocated) rather than `max_memory_reserved` 

You should observe that the behavior matches your expectations when you use `max_memory_allocated` (`max_memory_reserved` has to do with the memory reserved by our CUDACachingAllocator rather than the memory currently in use)

**Comment by mikaylagawarecki** [comment]
Closing as the behavior is expected, but feel free to comment if you have further thoughts

**Comment by threewayhandshake** [comment]
Thank you so much mikaylagawarecki for your response and helpful information!
The reason I used `max_memory_reserved` is simply because it matches `nvidia-smi`'s `Memory-Usage` value better than `max_memory_allocated`.

I confirmed that when using `max_memory_allocated`, there is no difference in max VRAM usage between variable and fixed length.
And I also confirmed that increased `max_memory_reserved` can be reduced back by `empty_cache()`.

If you don't mind, I want to ask a question at last.
If `max_memory_reserved` of a process is large but `max_memory_allocated` is small, will the VRAM available to other processes not be so limited?
Or it will be limited according to `max_memory_reserved`?

If if the latter, I will call `empty_cache()` moderately to prevent `max_memory_reserved` from becoming too large, but is there a more direct function that disables the cache mechanism or sets some sort of upper limit?

**Comment by threewayhandshake** [comment]
After asking the question, I searched for such a function myself, and thanks to you giving the keyword `CUDACachingAllocator`, I found several useful features.

As for "setting some sort of upper limit", it will probably help me to use `torch.cuda.set_per_process_memory_fraction()`.

Also `PYTORCH_CUDA_ALLOC_CONF=backend:cudaMallocAsync` prevents `max_memory_reserved` from getting large even without `set_per_process_memory_fraction()`, and seems to have a good performance than calling `empty_cache()` frequently.
